Sources/MongoSwift/SDAM.swift

```diff
@@ -24,11 +33,51 @@ public struct ServerAddress: Equatable {
 
     /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
     internal init(_ hostAndPort: String = "localhost:27017") throws {
-        let parts = hostAndPort.split(separator: ":")
-        self.host = String(parts[0])
-        guard let port = UInt16(parts[1]) else {
-            throw MongoError.InternalError(message: "couldn't parse address from \(hostAndPort)")
+        guard !hostAndPort.contains("?") else {
+            throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
         }
+
+        var host: String
+        var port: UInt16
+
+        // Check if host is an IPv6 literal.
+        if hostAndPort.first == "[" {
+            let ipLiteralRegex = try NSRegularExpression(pattern: #"^\[(.*)\](?::([0-9]+))?$"#)
+            guard
+                let match = ipLiteralRegex.firstMatch(
+                    in: hostAndPort,
+                    range: NSRange(location: 0, length: hostAndPort.utf16.count)
+                ),
+                let hostRange = Range(match.range(at: 1), in: hostAndPort)
+            else {
+                throw MongoError.InvalidArgumentError(message: "couldn't parse address from \(hostAndPort)")
+            }
+            host = String(hostAndPort[hostRange])
+            if let portRange = Range(match.range(at: 2), in: hostAndPort) {
+                port = UInt16(hostAndPort[portRange]) ?? 27017
```

Comment:
Similar to in the other branch, we need to be sure to throw errors here instead of defaulting to 27017 when `UInt16(hostAndPort[portRange])` returns nil. Since we'll be doing the same logic in both branches, we may want to consider defining a function for it to reduce code duplication.

Reply:
created an private static func verifyPort